src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/24.diff.sql

```diff
@@ -0,0 +1,146 @@
+--     @lastProcessedUtcDateTime
+--         * The last checkpoint datetime in UTC time (Coordinated Universal Time).
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE OR ALTER PROCEDURE dbo.FetchResourceChanges_3
+    @startId bigint,
+    @lastProcessedUtcDateTime datetime2(7),
+    @pageSize smallint
+AS
+BEGIN
+   
+    SET NOCOUNT ON;
+
+    /* Finds the prior partition to the current partition where the last processed watermark lies. It is a normal scenario when a prior watermark exists. */
+    DECLARE @precedingPartitionBoundary AS datetime2(7) = (SELECT TOP(1) CAST(prv.value as datetime2(7)) AS value FROM sys.partition_range_values AS prv WITH (NOLOCK)
+                                                            INNER JOIN sys.partition_functions AS pf WITH (NOLOCK) ON pf.function_id = prv.function_id
+                                                        WHERE pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+                                                            AND SQL_VARIANT_PROPERTY(prv.Value, 'BaseType') = 'datetime2'
+                                                            AND CAST(prv.value AS datetime2(7)) < DATEADD(HOUR, DATEDIFF(HOUR, 0, @lastProcessedUtcDateTime), 0)
+                                                        ORDER BY prv.boundary_id DESC);    
+
+    /* It ensures that it will not check resource changes in future partitions. */
+    DECLARE @endDateTimeToFilter AS datetime2(7) = DATEADD(HOUR, 1, SYSUTCDATETIME());
+
+    WITH PartitionBoundaries
+    AS (
+        SELECT CAST(prv.value as datetime2(7)) AS PartitionBoundary FROM sys.partition_range_values AS prv WITH (NOLOCK)
+            INNER JOIN sys.partition_functions AS pf WITH (NOLOCK) ON pf.function_id = prv.function_id
+        WHERE pf.name = N'PartitionFunction_ResourceChangeData_Timestamp'
+                AND SQL_VARIANT_PROPERTY(prv.Value, 'BaseType') = 'datetime2'
+                AND (
+                    /* Normal logic when prior watermark exists, Grab prior partition to ensure we do not miss data that was written across a partition boundary,
+                       and includes partitions until current one to ensure we keep moving to the next partition when some partitions do not have any resource changes. */
+                    CAST(prv.value AS datetime2(7)) BETWEEN @precedingPartitionBoundary AND @endDateTimeToFilter
+                    OR 
+                    /* It happens when no prior watermark exists or the last processed datetime of prior watermark is older than the last retention datetime.
+                       Uses the partition anchor datetime as the last processed DateTime. */
+                    @precedingPartitionBoundary IS NULL AND CAST(prv.value AS datetime2(7)) BETWEEN CONVERT(datetime2(7), N'1970-01-01T00:00:00.0000000') AND @endDateTimeToFilter
```

Comment:
Suggestion: It would be cleaner to have an ISNULL check on the precedingPartitionBoundary outside the query.  I.E. if precedingPartitionBoundary is null just set it to the the default.  This will remove the OR and make the query a little simpler and I think the procedure a little more readable.

Reply:
Sure, updated and checked the NULL logic outside the select query.